File: OnionMan_Unreal/Source/OnionMan_Unreal/Private/Network/NetworkManager.cpp
// Fill out your copyright notice in the Description page of Project Settings.


#include "Network/NetworkManager.h"
#include "Network/EncodingUtility.h"
#include "LogUtils.h"

using namespace OnionMan::Network;

UNetworkManager::UNetworkManager()
{
    LOG_ERROR("Creating NetworkManager")
}

void UNetworkManager::AddSynchronizedObject(ISynchronizedObjectBase& obj)
{
    uint32 objID = obj.ObjectID();
    if (m_synchronizedObjects.Contains(objID))
    {
        LOG_ERROR("The object ID %i is already used", objID);
        return;
    }
    m_synchronizedObjects.Add(objID, &obj);
    m_objectsArray.Add(&obj);
}

void UNetworkManager::RemoveSynchronizedObject(ISynchronizedObjectBase& obj)
{
    uint32 objID = obj.ObjectID();
    if (!m_synchronizedObjects.Contains(objID))
    {
        LOG_ERROR("No object with ID %i", objID);
        return;
    }
    m_synchronizedObjects.Remove(objID);
    m_objectsArray.Remove(&obj);
}

void UNetworkManager::GetObjectsToSync(TArray<TTuple<TObjectPtr<ISynchronizedObjectBase>, int>>& outObjectsToSync)
{
    outObjectsToSync.Empty();
    for (TObjectPtr<ISynchronizedObjectBase> synchronizedObject : m_objectsArray)
    {
        ObjectNeedSyncResult needSyncResult = synchronizedObject->NeedSync();
        if (needSyncResult.NeedSync())
        {
            outObjectsToSync.Add(TTuple<TObjectPtr<ISynchronizedObjectBase>, int>{synchronizedObject, needSyncResult.EncodedSize()});
        }
    }
}

void UNetworkManager::OnClientConnected(FString client)
{
    LOG_ERROR("New Client Connected, waw : %s", *client);
}

void UNetworkManager::OnBytesRecieved(UPARAM(ref) TArray<uint8>& encodedObjects)
{
    DecodeObjects(encodedObjects);
}

bool UNetworkManager::TryAcknoledge(UPARAM(ref) TArray<uint8>& recievedBytes, TArray<uint8>& outEncodedAck, FString& outIP)
{
    int offset = 0;
    outIP = EncodingUtility::Decode<FString>(recievedBytes, offset, recievedBytes.Num());

    TArray<FString> ip{};
    outIP.ParseIntoArray(ip, TEXT("."));
    LOG_ERROR("Recieved IP : %s", *outIP);

    if (ip.Num() != 4) 
    {
        return false;
    }

    for (FString stringByte : ip)
    {
        if (stringByte.IsNumeric())
        {
            int byte = FCString::Atoi(*stringByte);
            if (0 > byte || 255 < byte) 
            {
                return false;
            }
        }
        else 
        {
            return false;
        }
    }

    offset = 0;
    outEncodedAck.SetNumUninitialized(EncodingUtility::GetSizeOf<FString>(&AcknoledgeMessage));
    EncodingUtility::PutEncodedValueInBuffer<FString>(AcknoledgeMessage, outEncodedAck, offset);
    return true;
}

bool UNetworkManager::TryEncodeObjects(TArray<UNetworkBatch*>& outNetworkBatches)
{
    TArray<TTuple<TObjectPtr<ISynchronizedObjectBase>, int>> objectsToSync{};
    GetObjectsToSync(objectsToSync);

    if (objectsToSync.Num() == 0) 
    {
        outNetworkBatches = TArray<UNetworkBatch*>{};
        return false;
    }
    LOG_ERROR("%i objectsToSyncs : id : %i", objectsToSync.Num(), objectsToSync[0].Key->ObjectID())

    TArray<BatchInfos*> encodedObjectsBatches = TArray<BatchInfos*>();
    encodedObjectsBatches.Add(new BatchInfos(TArray<TObjectPtr<ISynchronizedObjectBase>>(), 0));
    int currentBatchIndex = 0;

    for(TTuple<TObjectPtr<ISynchronizedObjectBase>, int> object : objectsToSync)
    {
        BatchInfos* currentBatckInfos = encodedObjectsBatches[currentBatchIndex];
        if (currentBatckInfos->BatchSize + object.Value > MAX_BATCH_SIZE)
        {
            BatchInfos* newBatch = new BatchInfos(TArray<TObjectPtr<ISynchronizedObjectBase>>(), 0);
            newBatch->BatchSize = object.Value;
            newBatch->ObjectsToSync.Add(object.Key);
            encodedObjectsBatches.Add(newBatch);

            currentBatchIndex++;
        }
        else
        {
            currentBatckInfos->BatchSize += object.Value;
            currentBatckInfos->ObjectsToSync.Add(object.Key);
        }
    }

    outNetworkBatches = TArray<UNetworkBatch*>{};
    for(BatchInfos* batch : encodedObjectsBatches)
    {
        TArray<uint8> encodedObjectsBuffer = TArray<uint8>{};
        encodedObjectsBuffer.SetNumUninitialized(batch->BatchSize);
        int offset = 0;

        for (TObjectPtr<ISynchronizedObjectBase> synchronizedObject : batch->ObjectsToSync)
        {
            synchronizedObject->PutEncodedObjectToBuffer(encodedObjectsBuffer, offset, true);
        }

        UNetworkBatch* a = NewObject<UNetworkBatch>();
        a->EncodedBatch = encodedObjectsBuffer;
        outNetworkBatches.Add(a);

        delete batch;
    }

    return true;
}

void UNetworkManager::EncodeObjects(TArray<uint8>& outEncodedObjects)
{
    TArray<TTuple<TObjectPtr<ISynchronizedObjectBase>, int>> objectsToSync{};
    GetObjectsToSync(objectsToSync);

    int encodedObjectsTotalSize = 0;
    for (TTuple<TObjectPtr<ISynchronizedObjectBase>, int> obj : objectsToSync)
    {
        encodedObjectsTotalSize += obj.Value;
    }

    LOG_ERROR("Encoded Size = %i", encodedObjectsTotalSize);

    outEncodedObjects.Empty();
    outEncodedObjects.SetNumUninitialized(encodedObjectsTotalSize);
    int offset = 0;

    for (TTuple<TObjectPtr<ISynchronizedObjectBase>, int> obj : objectsToSync)
    {
        LOG_ERROR("Encoding object %i at offset %i", obj.Key->ObjectID(), offset);
        obj.Key->PutEncodedObjectToBuffer(outEncodedObjects, offset);
    }
}

void UNetworkManager::DecodeObjects(UPARAM(ref) TArray<uint8>& encodedObjects)
{
    int offset = 0;
    int bufferSize = encodedObjects.Num();
    while (offset < bufferSize)
    {
        int objectSize = EncodingUtility::Decode<int>(encodedObjects, offset);
        int initialOffset = offset;

        uint32 objectID = EncodingUtility::Decode<uint32>(encodedObjects, offset);
        int objectDataSize = objectSize - sizeof(uint32);

        if (m_synchronizedObjects.Contains(objectID))
        {
            TObjectPtr<ISynchronizedObjectBase> synchronizedObject = m_synchronizedObjects[objectID];
            synchronizedObject->DecodeObject(encodedObjects, offset, objectDataSize);
            if (offset - initialOffset != objectSize)
            {
                LOG_ERROR("Offset Overflow !")
            }
        }
        else
        {
            LOG_ERROR("There are no object with ID %i", objectID);
        }
    }
}

int UNetworkManager::GetMaxBatchSize()
{
    return MAX_BATCH_SIZE;
}

void UNetworkManager::TestMethod()
{
    /*TArray<TArray<int>&> testArray = TArray<TArray<int>&>{};

    for (int d = 0; d < 10; d++) 
    {
        TArray<int> dArray = TArray<int>{};

        for (int i = 0; i < 10; i++)
        {
            dArray.Add(d * 10 + i);
        }
        testArray.Add(dArray);
    }

    LOG_ERROR("Array of Array COUNT : %i", testArray.Num());
    for (int i = 0; i < testArray.Num(); i++) 
    {
        LOG_WARNING("Sub array %i : Count : %i", i, testArray[i].Num())
            
        for (int j = 0; j < testArray[i].Num(); j++)
        {
            LOG_WARNING("element at %i : Value : %i", j, testArray[i][j])

        }

    }*/
}

File: OnionMan_Unity/Assets/_Assets/Network/Encoding/NetworkManager.cs
using System.Collections.Generic;
using UnityEngine;
using System.IO;
using System.Linq;

using OnionMan.Utils;
using System.Net.Sockets;
using UnityEditor.Experimental.GraphView;
using UnityEngine.EventSystems;
using System.Threading;
using System;

namespace OnionMan.Network
{
    public class NetworkManager : SingletonMonoBehaviour<NetworkManager>
    {
        private class BatchInfos
        {
            public List<ISynchronizedObject> ObjectsToSync;
            public int BatchSize;

            public BatchInfos(List<ISynchronizedObject> objects, int batchSize)
            {
                ObjectsToSync = objects;
                BatchSize = batchSize;
            }
        }

        [SerializeField] private string m_ip;
        [SerializeField] private int m_port;

        private const int MAX_BATCH_SIZE = 8096;

        private Dictionary<uint, ISynchronizedObject> m_synchronizedObjects = new Dictionary<uint, ISynchronizedObject>();

        private TcpClient m_tcpClient;
        private NetworkStream m_networkStream;

        private int m_networkStreamOffset;

        Thread m_networkThread;

        private bool m_isConnected = false;
        private bool m_shouldListenForNetwork = false;

        public NetworkManager()
        {

        }

        protected virtual void LateUpdate()
        {
            if (m_isConnected)
            {
                if (TryEncodeObjects(out List<byte[]> encodedObjectsBatches))
                {
                    foreach (byte[] batch in encodedObjectsBatches)
                    {
                        m_networkStream.Write(batch);
                    }
                }
            }
        }

        public override void Initialize()
        {
            base.Initialize();
        }

        public void Connect()
        {
            if (!m_isConnected)
            {
                m_tcpClient = new TcpClient(m_ip, m_port);

                Debug.LogError($"Is Connected before connect: {m_tcpClient.Connected}");

                m_tcpClient.Connect(m_ip, m_port);

                Debug.LogError($"Is Connected after connect: {m_tcpClient.Connected}");

                m_networkStream = m_tcpClient.GetStream();
                m_isConnected = true;

                m_shouldListenForNetwork = true;
                m_networkThread = new Thread(new ThreadStart(ListenForNetwork));
            }
        }

        private void ListenForNetwork()
        {
            while (m_shouldListenForNetwork)
            {
                try
                {
                    byte[] encodedObjectsBuffer = new byte[MAX_BATCH_SIZE];
                    int encodedMessageSize = m_networkStream.Read(encodedObjectsBuffer, m_networkStreamOffset, MAX_BATCH_SIZE);

                    byte[] resizedEncodedObjects = new byte[encodedMessageSize];
                    EncodingUtility.ResizeBuffer(encodedObjectsBuffer, ref resizedEncodedObjects);

                    DecodeObjects(resizedEncodedObjects);
                }
                catch (ArgumentNullException e)
                {
                    Debug.LogException(e);
                }
                catch (SocketException e)
                {
                    Debug.LogException(e);
                }
                catch(Exception e)
                {
                    Debug.LogException(e);
                }
            }
        }

        public void AddSynchronizedObject(ISynchronizedObject obj)
        {
            uint objID = obj.ObjectID;
            if (m_synchronizedObjects.ContainsKey(objID))
            {
                throw new InvalidDataException($"The ID {objID} is already used by {m_synchronizedObjects[objID]}");
            }
            m_synchronizedObjects.Add(objID, obj);
        }

        public void RemoveSynchronizedObject(ISynchronizedObject obj)
        {
            uint objID = obj.ObjectID;
            if (!m_synchronizedObjects.ContainsKey(objID))
            {
                throw new InvalidDataException($"No object with ID {objID}");
            }
            m_synchronizedObjects.Remove(objID);
        }

        public IEnumerable<(ISynchronizedObject, int)> GetObjectsToSync()
        {
            foreach (ISynchronizedObject synchronizedObject in m_synchronizedObjects.Values)
            {
                ObjectNeedSyncResult needSyncResult = synchronizedObject.NeedSync();
                if (needSyncResult.NeedSync)
                {
                    yield return (synchronizedObject, needSyncResult.EncodedSize);
                }
            }
        }

        public bool TryEncodeObjects(out List<byte[]> encodedObjectsBufferBatch)
        {
            IEnumerable<(ISynchronizedObject, int)> objectsToSync = GetObjectsToSync();

            if (objectsToSync.Count() == 0)
            {
                encodedObjectsBufferBatch = new List<byte[]>();
                return false;
            }

            List<BatchInfos> encodedObjectsBatches = new List<BatchInfos>()
            {
                new BatchInfos(new List<ISynchronizedObject>(), 0)
            };
            int currentBatchIndex = 0;
            foreach((ISynchronizedObject synchronizedObject, int encodedSize) in objectsToSync)
            {
                BatchInfos currentBatckInfos = encodedObjectsBatches[currentBatchIndex];
                if (currentBatckInfos.BatchSize + encodedSize > MAX_BATCH_SIZE)
                {
                    encodedObjectsBatches.Add(new BatchInfos(
                        new List<ISynchronizedObject>(){ synchronizedObject },
                        encodedSize));

                    currentBatchIndex++;
                }
                else
                {
                    currentBatckInfos.BatchSize += encodedSize;
                    currentBatckInfos.ObjectsToSync.Add(synchronizedObject);
                }
            }

            encodedObjectsBufferBatch = new List<byte[]>();
            foreach (BatchInfos batch in encodedObjectsBatches)
            {
                byte[] encodedObjectsBuffer = new byte[batch.BatchSize];
                int offset = 0;

                foreach (ISynchronizedObject synchronizedObject in batch.ObjectsToSync)
                {
                    synchronizedObject.PutEncodedObjectToBuffer(ref encodedObjectsBuffer, ref offset);
                }
                encodedObjectsBufferBatch.Add(encodedObjectsBuffer);
            }

            return true;
        }

        public void DecodeObjects(byte[] encodedObjects)
        {
            int offset = 0;
            while (offset < encodedObjects.Length)
            {
                int objectSize = EncodingUtility.Decode<int>(encodedObjects, ref offset);
                int initialOffset = offset;

                uint objectID = EncodingUtility.Decode<uint>(encodedObjects, ref offset);
                int objectDataSize = objectSize - sizeof(uint);

                if (m_synchronizedObjects.TryGetValue(objectID, out ISynchronizedObject synchronizedObject))
                {
                    synchronizedObject.DecodeObject(encodedObjects, ref offset, objectDataSize);
                    if (offset - initialOffset != objectSize)
                    {
                        Debug.LogError("Offset Overflow !");
                    }
                }
                else
                {
                    throw new InvalidDataException($"There are no object with ID {objectID}");
                }
            }
        }
    }
}

